Memory leak on test run

# Scope
 Tempesta: Commit 930ba7dc761f8236424894fddc40cd5cc1c7af7f
 Tests Commit: df8a0e7ec6f4a4cec96666db9f953490a710c564
  Test: `./run_tests.py -n t_stress.test_wordpress.H2WordpressStress.test_get_large_images`

Every test run we lose a memory. I can't reproduce it without the test.

The "docker build" failed:

```
 ---> 8b951aad8184
Step 19/28 : RUN wp core install --path="${WORDPRESS_DIR}" --url="${WORDPRESS_URL}" --title="Tempesta Wordpress Test" --admin_user=admin --admin_password=secret --admin_email=test@tempesta-tech.com --skip-email
 ---> Running in 2d6ff0bb3064
Error: Error: WordPress 6.6.1 requires MySQL 5.5.5 or higher
```

```
======================================================================
ERROR: test_get_large_images (t_stress.test_wordpress.H2WordpressStress)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kingluo/tempesta-test/helpers/remote.py", line 107, in run_cmd
    assert p.returncode == 0, "Cmd: '%s' return code is not 0 (%d)." % (
AssertionError: Cmd: 'cat /tmp/nginx/tt-wordpress.tar.gz | docker build -  --tag tt-wordpress --network host' return code is not 0 (1).

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kingluo/tempesta-test/t_stress/test_wordpress.py", line 74, in test_get_large_images
    self.start_all()
  File "/home/kingluo/tempesta-test/t_stress/test_wordpress.py", line 69, in start_all
    self.start_all_servers()
  File "/home/kingluo/tempesta-test/framework/tester.py", line 320, in start_all_servers
    srv.start()
  File "/home/kingluo/tempesta-test/helpers/stateful.py", line 60, in start
    self.run_start()
  File "/home/kingluo/tempesta-test/framework/docker_server.py", line 123, in run_start
    self._build_image()
  File "/home/kingluo/tempesta-test/framework/docker_server.py", line 168, in _build_image
    stdout, stderr = self.node.run_cmd(
  File "/home/kingluo/tempesta-test/helpers/remote.py", line 121, in run_cmd
    raise CmdError(
helpers.remote.CmdError: ('Сmd cat /tmp/nginx/tt-wordpress.tar.gz | docker build -  --tag tt-wordpress --network host exited with return code 1', b'', b'#0 building with
"default" instance using docker driver\n\n#1 [internal] load remote build context\n#1 DONE 0.0s\n\n#2 copy /context /\n#2 DONE 0.0s\n\n#3 [internal] load metadata for doc
ker.io/library/wordpress:php8.1-apache\n#3 DONE 1.1s\n\n#4 [internal] load metadata for docker.io/composer/composer:2.4.3\n#4 DONE 2.1s\n\n#5 FROM docker.io/composer/comp
oser:2.4.3@sha256:2d5ee5c7dedd179206563d6631b6cf6ca0bca9561dd728de4de717e5ce0b63a2\n#5 DONE 0.0s\n\n#6 [stage-0  1/17] FROM docker.io/library/wordpress:php8.1-apache@sha2
56:aa78ee76c622598ea8e7807f1c391888e4ee60ac8f9296771a943515a25b2a04\n#6 DONE 0.0s\n\n#7 [stage-0  6/17] COPY --chown=www-data composer.json /var/www/.composer/composer.js
on\n#7 CACHED\n\n#8 [stage-0  7/17] COPY tt-docker-entrypoint.sh /usr/local/bin\n#8 CACHED\n\n#9 [stage-0  8/17] RUN chmod +x /usr/local/bin/tt-docker-entrypoint.sh\n#9 C
ACHED\n\n#10 [stage-0  2/17] COPY --from=composer/composer:2.4.3 /usr/bin/composer /usr/bin/composer\n#10 CACHED\n\n#11 [stage-0  4/17] RUN chown www-data /var/www\n#11 C
ACHED\n\n#12 [stage-0  3/17] RUN apt-get -y update && apt install -y libzip-dev imagemagick\n#12 CACHED\n\n#13 [stage-0  9/17] RUN composer global update --no-interaction
\n#13 CACHED\n\n#14 [stage-0  5/17] COPY --chown=www-data wp-config.php /usr/src/wordpress/wp-config.php\n#14 CACHED\n\n#15 [stage-0 10/17] RUN cp /var/www/.composer/wp-c
ontent/wp-sqlite-db/src/db.php "/usr/src/wordpress/wp-content/db.php"\n#15 CACHED\n\n#16 [stage-0 11/17] RUN wp core install --path="/usr/src/wordpress" --url="http://127
.0.0.1" --title="Tempesta Wordpress Test" --admin_user=admin --admin_password=secret --admin_email=test@tempesta-tech.com --skip-email\n#16 0.317 Error: Error: WordPress
6.6.1 requires MySQL 5.5.5 or higher\n#16 ERROR: process "/bin/sh -c wp core install --path=\\"${WORDPRESS_DIR}\\" --url=\\"${WORDPRESS_URL}\\" --title=\\"Tempesta Wordpr
ess Test\\" --admin_user=admin --admin_password=secret --admin_email=test@tempesta-tech.com --skip-email" did not complete successfully: exit code: 1\n------\n > [stage-0
 11/17] RUN wp core install --path="/usr/src/wordpress" --url="http://127.0.0.1" --title="Tempesta Wordpress Test" --admin_user=admin --admin_password=secret --admin_emai
l=test@tempesta-tech.com --skip-email:\n0.317 Error: Error: WordPress 6.6.1 requires MySQL 5.5.5 or higher\n------\nDockerfile:27\n--------------------\n  25 |     RUN cp
 /var/www/.composer/wp-content/wp-sqlite-db/src/db.php "${WORDPRESS_DIR}/wp-content/db.php"\n  26 |     \n  27 | >>> RUN wp core install --path="${WORDPRESS_DIR}" --url="
${WORDPRESS_URL}" --title="Tempesta Wordpress Test" --admin_user=admin --admin_password=secret --admin_email=test@tempesta-tech.com --skip-email\n  28 |     \n  29 |
# Generate images\n--------------------\nERROR: failed to solve: process "/bin/sh -c wp core install --path=\\"${WORDPRESS_DIR}\\" --url=\\"${WORDPRESS_URL}\\" --title=\\
"Tempesta Wordpress Test\\" --admin_user=admin --admin_password=secret --admin_email=test@tempesta-tech.com --skip-email" did not complete successfully: exit code: 1\n',
1)
```

--- a/fw/cache.c
+++ b/fw/cache.c
@@ -2703,17 +2703,12 @@ tfw_cache_build_resp_body(TDB *db, TdbVRec *trec, TfwMsgIter *it, char *p,
 	/*
 	 * If all skbs/frags are used up (see @tfw_http_msg_expand_data()),
 	 * create new skb with empty frags to reference the cached body;
-	 * otherwise, use next empty frag in current skb. Create a new skb, if
-	 * TX flags for headers and body differ.
+	 * otherwise, use next empty frag in current skb.
 	 */
-	if (!it->skb || (it->frag + 1 >= MAX_SKB_FRAGS)
-	    || (sh_frag == !(skb_shinfo(it->skb)->tx_flags & SKBTX_SHARED_FRAG)))
-	{
+	if (!it->skb || it->frag + 1 >= MAX_SKB_FRAGS || sh_frag) {
 		if  ((r = tfw_msg_iter_append_skb(it)))
 			return r;
-		if (!sh_frag)
-			skb_shinfo(it->skb)->tx_flags &= ~SKBTX_SHARED_FRAG;
-		else
+		if (sh_frag)
 			skb_shinfo(it->skb)->tx_flags |= SKBTX_SHARED_FRAG;
 	}
 

--- a/fw/http.c
+++ b/fw/http.c
@@ -4838,12 +4838,9 @@ tfw_h2_append_predefined_body(TfwHttpResp *resp, const TfwStr *body)
 		return -EINVAL;
 	it->frag = skb_shinfo(it->skb)->nr_frags - 1;
 
-	if ((it->frag + 1 >= MAX_SKB_FRAGS)
-	    || (skb_shinfo(it->skb)->tx_flags & SKBTX_SHARED_FRAG))
-	{
-		if  ((r = tfw_msg_iter_append_skb(it)))
+	if (it->frag + 1 >= MAX_SKB_FRAGS) {
+		if ((r = tfw_msg_iter_append_skb(it)))
 			return r;
-		skb_shinfo(it->skb)->tx_flags &= ~SKBTX_SHARED_FRAG;
 	}
 
 	data = body->data;

--- a/fw/ss_skb.c
+++ b/fw/ss_skb.c
@@ -1593,6 +1593,7 @@ ss_skb_to_sgvec_with_new_pages(struct sk_buff *skb, struct scatterlist *sgl,
 				int order = get_order(size);
 				gfp_t flags = order > 0 ?
 					GFP_ATOMIC | __GFP_COMP : GFP_ATOMIC;
+
 				p = alloc_pages(flags, order);
 				if (!p)
 					return -ENOMEM;

--- a/fw/ss_skb.c
+++ b/fw/ss_skb.c
@@ -1591,7 +1591,9 @@ ss_skb_to_sgvec_with_new_pages(struct sk_buff *skb, struct scatterlist *sgl,
 			size = skb_frag_size(f);
 			if (remain < size) {
 				int order = get_order(size);
-				p = alloc_pages(GFP_ATOMIC, order);
+				gfp_t flags = order > 0 ?
+					GFP_ATOMIC | __GFP_COMP : GFP_ATOMIC;
+				p = alloc_pages(flags, order);
 				if (!p)
 					return -ENOMEM;
 				remain = (1 << order) * PAGE_SIZE;

--- a/tls/ttls.c
+++ b/tls/ttls.c
@@ -1502,7 +1502,10 @@ ttls_parse_certificate(TlsCtx *tls, unsigned char *buf, size_t len,
 	 * TODO call ttls_update_checksum() for the message as well.
 	 */
 	T_FSM_STATE(TTLS_CC_HS_ALLOC) {
-		if (!(pg = alloc_pages(GFP_ATOMIC, get_order(io->hslen)))) {
+		int order = get_order(io->hslen);
+		gfp_t flags = order > 0 ? GFP_ATOMIC | __GFP_COMP : GFP_ATOMIC;
+
+		if (!(pg = alloc_pages(flags, order))) {
 			T_WARN("TLS: cannot allocate pages for a certificate\n");
 			return -ENOMEM;
 		}

